virtual_rainforest/core/grid.py

```diff
@@ -446,3 +452,132 @@ def populate_distances(
         """
 
         self._distances = squareform(pdist(self.centroids))
+
+    def map_xy_to_cell_id(
+        self,
+        x_coords: np.ndarray,
+        y_coords: np.ndarray,
+    ) -> list[list[Optional[int]]]:
+        """Map a set of coordinates onto grid cells.
```

Comment:
Unless I'm misunderstanding something about the problem -- which is entirely possible -- I'm *really* concerned about this as a design. It makes sense that you'd want to know which cell, if any, some coordinates fall within, but do you really need to know whether a point is on the intersection of multiple cells? This will make life more complicated for calling code and I can't see what the gain would be.

Why don't you just have the boundaries of your cells be thresholds, so that coordinates always fall on one side of them or the other? The problem with having a given coordinate potentially fall within multiple cells is that it then becomes difficult to decide whether, say, a given tree is in cell A or cell B. In practice, of course, basically all your fauna and flora etc. are going to end up being in one cell or another (or none), so you might as well make this an assumption of your models. The alternative is that callers of this function have to deal with this complexity which will lead to ugly, finicky code everywhere.

Reply:
Oh absolutely. That would be mad! This function is purely used as a filter to detect when cell coordinates do not map uniquely onto a single cell. The science modules are not planned to have spatial structure _within_ cells, so this is only used for checking data arrays align uniquely onto cells. Using the `intersects` method is not sensitive to the shape of the cell.

Reply:
If you are only using this function to check the validity of the input arguments I'd just do that then, rather than returning a more complex data structure.